File: Cpp/pkg/AmoreCpp/src/SimpleNeuron.cpp
/*
 * SimpleNeuron.cpp
 *
 *  Created on: 25/05/2011
 *      Author: mcasl
 */



#include "package.h"
#include "classHeaders/NeuralFactory.h"
#include "classHeaders/Container.h"
#include "classHeaders/Iterator.h"
#include "classHeaders/ActivationFunction.h"
#include "classHeaders/PredictBehavior.h"
#include "classHeaders/SimpleNeuron.h"

SimpleNeuron::SimpleNeuron(NeuralFactory& neuralFactory) :
  Neuron(neuralFactory)
{
}


double
SimpleNeuron::getInducedLocalField()
{
  return d_inducedLocalField;
}

void
SimpleNeuron::setInducedLocalField(double inducedLocalField)
{
  d_inducedLocalField = inducedLocalField;
}

double
SimpleNeuron::getOutput()
{
  return d_output;
}

void
SimpleNeuron::setOutput(double output)
{
  d_output = output;
}



double
SimpleNeuron::getTarget()
{
  return d_target;
}

void
SimpleNeuron::setTarget(double target)
{
  d_target = target;
}


double
SimpleNeuron::getOutputDerivative()
{
  return d_outputDerivative ;
}



void
SimpleNeuron::setOutputDerivative(double outputDerivative)
{
  d_outputDerivative = outputDerivative;
}


Handler
SimpleNeuron::getId()
{
  return d_Id;
}


void
SimpleNeuron::setId(Handler Id)
{
  d_Id = Id;
}

ConIteratorPtr
SimpleNeuron::getConIterator()
{
  return d_nCons->createIterator();
}

void
SimpleNeuron::addCon(ConPtr conPtr)
{
  d_nCons->push_back(conPtr);
}


void
SimpleNeuron::setActivationFunction(ActivationFunctionPtr activationFunctionPtr)
{
  d_activationFunction = activationFunctionPtr;
}


void
SimpleNeuron::setPredictBehavior(PredictBehaviorPtr predictBehaviorPtr)
{
  d_predictBehavior = predictBehaviorPtr;
}



void
SimpleNeuron::setNeuronTrainBehavior(NeuronTrainBehaviorPtr neuronTrainBehaviorPtr)
{
  d_neuronTrainBehavior = neuronTrainBehaviorPtr;
}



std::string
SimpleNeuron::getNeuronTrainBehaviorName()
{
  return d_neuronTrainBehavior->getName();
}


NeuralNetworkPtr
SimpleNeuron::getNeuralNetwork( )
{
  NeuralNetworkPtr neuralNetworkPtr ( d_neuralNetwork.lock() );
  return neuralNetworkPtr;
}


void
SimpleNeuron::setNeuralNetwork( NeuralNetworkPtr neuralNetworkPtr)
{
  d_neuralNetwork = neuralNetworkPtr;
}



double
SimpleNeuron::useActivationFunctionf0()
{
  return d_activationFunction->f0();
}


double
SimpleNeuron::useActivationFunctionf1()
{
  return d_activationFunction->f1();
}



void
SimpleNeuron::singlePatternForwardAction()
{
  d_predictBehavior->singlePatternForwardAction();
}


void
SimpleNeuron::singlePatternBackwardAction()
{
  d_neuronTrainBehavior->singlePatternBackwardAction();
}


void
SimpleNeuron::show()
{
  if (d_nCons->size() == 0)
    {

      int id = getId();
      Rprintf("\n\n-----------------------------------");
      if (id == NA_INTEGER)
        {
          Rprintf("\n Id: NA, Invalid neuron Id");
        }
      else
        {
          Rprintf("\n Id: %d", id);
        }
      Rprintf("\n-----------------------------------");
      Rprintf("\n output: %lf", d_output);
      Rprintf("\n-----------------------------------");
      Rprintf("\n Neuron Train Behavior: %s", getNeuronTrainBehaviorName().c_str() );
      Rprintf("\n-----------------------------------");

    }
  else
    {
      int id = getId();
      Rprintf("\n\n-----------------------------------");
      if (id == NA_INTEGER)
        {
          Rprintf("\n Id: NA, Invalid neuron Id");
        }
      else
        {
          Rprintf("\n Id: %d", id);
        }
      Rprintf("\n-----------------------------------");
      d_predictBehavior->show();
      Rprintf("\n output: %lf", d_output);
      Rprintf("\n target: %lf", d_target);
      Rprintf("\n-----------------------------------");
      d_nCons->show();
      Rprintf("\n-----------------------------------");
      Rprintf("\n Neuron Train Behavior: %s", getNeuronTrainBehaviorName().c_str() );
      Rprintf("\n-----------------------------------");
    }
}

bool
SimpleNeuron::validate()
{
  BEGIN_RCPP
  if (getId() == NA_INTEGER ) throw std::range_error("[C++ SimpleNeuron::validate]: Error, Id is NA.");
    d_nCons->validate();
  return (TRUE);
END_RCPP}


int
SimpleNeuron::numberOfConnections()
{
  return static_cast<int>(d_nCons->size());
}



double
SimpleNeuron::costFunctionf0(double output, double target)
{
  NeuralNetworkPtr neuralNetworkPtr( d_neuralNetwork.lock() );
  return neuralNetworkPtr->costFunctionf0( output, target );
}


double
SimpleNeuron::costFunctionf1(double output, double target)
{
  NeuralNetworkPtr neuralNetworkPtr( d_neuralNetwork.lock() );
  return neuralNetworkPtr->costFunctionf1( output, target );
}



void
SimpleNeuron::addToBias(double value)
{
  d_predictBehavior->addToBias(value);
}


void
SimpleNeuron::addToDelta(double value)
{
  d_neuronTrainBehavior->addToDelta(value);
}

void
SimpleNeuron::setLearningRate(double learningRate)
{
  d_neuronTrainBehavior->setLearningRate(learningRate);
}

File: pkg/Java/AmoreJ/javasrc/es/unileon/amorej/SimpleNeuron.java
package es.unileon.amorej;

import es.unileon.amorej.net.NeuralNetwork;

public class SimpleNeuron implements Neuron {

	private String id;
	private double inducedLocalField;
	private double output;
	private double outputDerivative;
	private double target;
	private Container<Connection> connectionContainer;

	private NeuralNetwork neuralNetwork;

	// TODO PredictBehavior predictBehavior;
	// TODO ActivationFunction activationFunction;
	// TODO NeuronTrainBehavior neuronTrainBehavior;

	public SimpleNeuron(NeuralNetwork neuralNetwork) {
		this.neuralNetwork = neuralNetwork;
		this.id = "NA";
		this.inducedLocalField = 0.0;
		this.output = 0.0;
		this.outputDerivative = 0.0;
		this.target = 0.0;
		this.connectionContainer = neuralNetwork.getFactory().makeConnectionContainer();
	}

	public double getInducedLocalField() {
		return this.inducedLocalField;
	}

	public void setInducedLocalField(double inducedLocalField) {
		this.inducedLocalField = inducedLocalField;
	}

	public double getOutput() {
		return this.output;
	}

	public void setOutput(double output) {
		this.output = output;
	}

	public double getTarget() {
		return this.target;
	}

	public void setTarget(double target) {
		this.target = target;
	}

	public double getOutputDerivative() {
		return this.outputDerivative;
	}

	public void setOutputDerivative(double outputDerivative) {
		this.outputDerivative = outputDerivative;
	}

	public String getId() {
		return this.id;
	}

	public void setId(String Id) {
		this.id = Id;
	}

	public boolean validate() throws AmoreJException {
		if (this.id.equals("NA"))
			throw new AmoreJException(
					"[Java SimpleNeuron.validate]: Error, Id is NA.");
		if (this.id.equals(""))
			throw new AmoreJException(
					"[Java SimpleNeuron.validate]: Error, Id is empty string.");

		// Iterator<Connection> connectionItr =
		// connectionContainer.createIterator();
		// for (connectionItr.first(); !connectionItr.isDone();
		// connectionItr.next()) {
		//
		// }

		for (Connection connection : connectionContainer) {
			connection.validate();
		}
		return true;
	}

	public void addConnection(Connection connection) {
		connectionContainer.add(connection);
	}

	// TODO public void
	// setActivationFunction(ActivationFunction activationFunction)
	// {
	// this.activationFunction = activationFunction;
	// }
	//

	// TODO public void
	// setPredictBehavior(PredictBehavior predictBehavior)
	// {
	// this.predictBehavior = predictBehavior;
	// }
	//
	//
	//
	// TODO public void
	// setNeuronTrainBehavior(NeuronTrainBehavior neuronTrainBehavior)
	// {
	// this.neuronTrainBehavior = neuronTrainBehavior;
	// }
	//
	//
	//
	// TODO public String
	// getNeuronTrainBehaviorName()
	// {
	// return this.neuronTrainBehavior.getName();
	// }
	//
	//
	public NeuralNetwork getNeuralNetwork() {
		return this.neuralNetwork;
	}

	public void setNeuralNetwork(NeuralNetwork neuralNetwork) {
		this.neuralNetwork = neuralNetwork;
	}

	//
	//
	//
	// TODO public double
	// useActivationFunctionf0()
	// {
	// return this.activationFunction.f0();
	// }
	//
	//
	// TODO public double
	// useActivationFunctionf1()
	// {
	// return this.activationFunction.f1();
	// }
	//
	//
	//
	// TODO public void
	// singlePatternForwardAction()
	// {
	// this.predictBehavior.singlePatternForwardAction();
	// }
	//
	//
	// TODO public void
	// singlePatternBackwardAction()
	// {
	// this.neuronTrainBehavior.singlePatternBackwardAction();
	// }
	//

	public int numberOfConnections() {
		return connectionContainer.size();
	}

	public void show() {
		System.out.println("-----------------------------------");
		System.out.println(" id: " + this.id);
		System.out.println("-----------------------------------");
		// predictBehavior.show();
		System.out.println(" output: " + this.output);
		System.out.println(" target: " + this.target);
		System.out.print("-----------------------------------");
		for (Connection connection : this.connectionContainer) {
			connection.show();
		}
		System.out.println("\n-----------------------------------");
		// System.out.println(" Neuron Train Behavior: %s",
		// getNeuronTrainBehaviorName().c_str() );
		System.out.println("\n-----------------------------------");
	}
	// TODO public double costFunctionf0(double output, double target)
	// {
	// return this.neuralNetwork.costFunctionf0( output, target );
	// }
	//
	//
	// TODO public double costFunctionf1(double output, double target)
	// {
	// return this.neuralNetwork.costFunctionf1( output, target );
	// }
	//
	//
	// TODO public void addToBias(double value)
	// {
	// this.predictBehavior.addToBias(value);
	// }
	//
	//
	// TODO public void addToDelta(double value)
	// {
	// this.neuronTrainBehavior.addToDelta(value);
	// }
	//
	// TODO public void
	// setLearningRate(double learningRate)
	// {
	// this.neuronTrainBehavior.setLearningRate(learningRate);
	// }

}